src/platforms/atomic-kms/server/kms/display_buffer.cpp

```diff
@@ -335,6 +344,136 @@ void mga::DisplaySink::set_next_image(std::unique_ptr<Framebuffer> content)
     }
 }
 
+namespace {
+    auto const MAX_PLANES = 4zu;
+
+    auto get_import_buffers(std::shared_ptr<mg::DMABufBuffer> buffer)
+        -> std::tuple<std::array<int, 4>, std::array<int, 4>, std::array<int, 4>, std::array<uint64_t, 4>>
+    {
+        auto const plane_descriptors = buffer->planes();
+        assert(plane_descriptors.size() <= MAX_PLANES);
+
+        // std::array becuase we can't really return int[4]
+        std::array<int, 4> dmabuf_fds = {0};
+        std::array<int, 4> pitches = {0};
+        std::array<int, 4> offsets = {0};
+        std::array<uint64_t, 4> modifiers = {0};
+
+        for (std::size_t i = 0; i < std::min(MAX_PLANES, plane_descriptors.size()); i++)
+        {
+            dmabuf_fds[i] = plane_descriptors[i].dma_buf;
+            pitches[i] = plane_descriptors[i].stride;
+            offsets[i] = plane_descriptors[i].offset;
+            modifiers[i] = buffer->modifier().value_or(DRM_FORMAT_MOD_INVALID);
+        }
+
+        return std::make_tuple(dmabuf_fds, pitches, offsets, modifiers);
+    }
+
+    auto import_gbm_bo(
+        std::shared_ptr<struct gbm_device> gbm,
+        std::shared_ptr<mg::DMABufBuffer> buffer,
+        std::array<int, 4> dmabuf_fds,
+        std::array<int, 4> pitches,
+        std::array<int, 4> offsets) -> struct gbm_bo*
+    {
+        auto const plane_descriptors = buffer->planes();
+
+        gbm_import_fd_modifier_data import_data = {
+            .width = buffer->size().width.as_uint32_t(),
+            .height = buffer->size().height.as_uint32_t(),
+            .format = buffer->format(),
+            .num_fds = static_cast<uint32_t>(std::min(plane_descriptors.size(), MAX_PLANES)),
+            .fds = {dmabuf_fds[0], dmabuf_fds[1], dmabuf_fds[2], dmabuf_fds[3]},
+            .strides = {pitches[0], pitches[1], pitches[2], pitches[3]},
+            .offsets = {offsets[0], offsets[1], offsets[2], offsets[3]},
+            .modifier = buffer->modifier().value_or(DRM_FORMAT_MOD_NONE),
+        };
+
+        return gbm_bo_import(gbm.get(), GBM_BO_IMPORT_FD_MODIFIER, (void*)&import_data, GBM_BO_USE_SCANOUT);
+    }
+
+    auto drm_fb_id_from_dma_buffer(mir::Fd drm_fd, std::shared_ptr<struct gbm_device> const gbm, std::shared_ptr<mg::DMABufBuffer> buffer)
+        -> std::shared_ptr<uint32_t>
+    {
+
+        auto [dmabuf_fds, pitches, offsets, modifiers] = get_import_buffers(buffer);
+        auto* gbm_bo = import_gbm_bo(gbm, buffer, dmabuf_fds, pitches, offsets);
+        if (!gbm_bo)
+        {
+            mir::log_warning("Failed to import buffer");
+            return {};
+        }
+
+        auto const plane_descriptors = buffer->planes();
+        uint32_t gem_handles[MAX_PLANES] = {0};
+        for (std::size_t i = 0; i < std::min(MAX_PLANES, plane_descriptors.size()); i++)
+            gem_handles[i] = gbm_bo_get_handle_for_plane(gbm_bo, i).u32;
+
+        auto fb_id = std::shared_ptr<uint32_t>{
```

Comment:
As it stands, I think this code leaks a `gbm_bo` each time it's called? It calls `gbm_bo_import`, which will allocate a `gbm_bo`, but nothing every frees it.

Maybe `import_gbm_bo` should return a `shared_ptr<gbm_bo>` with cleanup functor, and that should in turn be captured by *this* `shared_ptr`?

Reply:
Hmm, I didn't know that, I thought we only got a (cached) handle to the buffer object (i.e. created once on the first import and reused later).

Reply:
Also, I've noticed that we import and destroy a buffer object every frame, this doesn't feel correct. I think it might also be related to the artifacts showing in glmark2